### PR TITLE
Enable backport of unmerged PRs

### PR DESCRIPTION
**Pull request by gnossen**
This PR enabled the backport of https://github.com/grpc/grpc/pull/25493. It adds two features:

 - The ability to backport unmerged PRs (including those with multiple commits)
 - The ability to run a script after cherry-picking the desired commits, including scripts that generate files.

It also fixes a bug where the script previously created PRs on the `grpc/grpc` repo instead of on the user's fork.

### CODE REVIEW

File: tools/release/backport_pr.sh [hunk 1]
```diff
@@ -24,59 +24,100 @@ ensure_command () {
   fi
 }
 
+display_usage () {
+  echo "USAGE: $0 PR_NUMBER GITHUB_USER BACKPORT_BRANCHES REVIEWERS [-c PER_BACKPORT_COMMAND]" >/dev/stderr
+  echo "   PR_NUMBER: The number for the PR to be backported." >/dev/stderr
+  echo "   GITHUB_USER: Your GitHub username." >/dev/stderr
+  echo "   BACKPORT_BRANCHES: A space-separated list of branches to which to backport." >/dev/stderr
```
Comment (drfloob) [inline tools/release/backport_pr.sh:31]: I'm not a language lawyer, so while I _think_ this is grammatically correct, it's a bit hard to parse. This is the list of branches that will receive the backported PR, yes?

File: tools/release/backport_pr.sh [hunk 2]
```diff
@@ -24,59 +24,100 @@ ensure_command () {
...
+  echo "   REVIEWERS: A comma-separated list of users add as reviewers and assignees." >/dev/stderr
```
Comment (drfloob) [inline tools/release/backport_pr.sh:32]: Missed a word. "... A comma-separated list of users *to* add as ..."

File: tools/release/backport_pr.sh [hunk 3]
```diff
@@ -24,59 +24,100 @@ ensure_command () {
   fi
 }
 
+display_usage () {
```
Comment (drfloob) [inline tools/release/backport_pr.sh:27]: Nit: Have you considered using a single expression for this? Either echoing a single multiline string to stderr or `cat > /dev/stderr -<< EOF` should work.

File: tools/release/backport_pr.sh [hunk 4]
```diff
@@ -24,59 +24,100 @@ ensure_command () {
...
+      display_usage
+      ;;
+    : )
+      echo "Invalid option: $OPTARG requires an argument." >/dev/stderr
+      display_usage
+      ;;
+  esac
+done
+
+if [[ ! -z "$(git status --porcelain)" && ! -z "$PER_BACKPORT_COMMAND" ]]; then
```
Comment (drfloob) [inline tools/release/backport_pr.sh:74]: I'm not sure I understand the logic here. What is the expected behavior of the script if the working dir is not clean, and the user _did not_ provide a per-backport command?

File: tools/release/backport_pr.sh [hunk 5]
```diff
@@ -24,59 +24,100 @@ ensure_command () {
...
-
-REVIEWERS="$4"
 
 PR_DATA=$(curl -s -u "$GITHUB_USER:$GITHUB_TOKEN" \
           -H "Accept: application/vnd.github.v3+json" \
           "https://api.github.com/repos/grpc/grpc/pulls/$PR_NUMBER")
 
-MERGE_COMMIT=$(echo "$PR_DATA" | jq -r '.merge_commit_sha')
+STATE=$(echo "$PR_DATA" | jq -r '.state')
+if [ "$STATE" == "merged" ]; then
```
Comment (drfloob) [inline tools/release/backport_pr.sh:98]: When is the state attribute set to "merged"? It looks like GitHub lets you filter on state being "open", "closed", or "all" in the PR list API. https://docs.github.com/en/rest/reference/pulls#list-pull-requests

**Review by drfloob — COMMENTED** [review]
This looks pretty handy! I have two questions on the logic here, and otherwise a couple nits.

File: tools/release/backport_pr.sh (hunk 1 above)
Comment (gnossen) [reply tools/release/backport_pr.sh]: Fair enough. I was a bit skeptical of it when I gave it a second look. I've gone back through the help text with an eye toward wordsmithing. Hopefully it's a bit more understandable now.

File: tools/release/backport_pr.sh (hunk 3 above)
Comment (gnossen) [reply tools/release/backport_pr.sh:27]: Much better stylistically. Switched.

File: tools/release/backport_pr.sh (hunk 4 above)
Comment (gnossen) [reply tools/release/backport_pr.sh:77]: > What is the expected behavior of the script if the working dir is not clean, and the user did not provide a per-backport command?

In that case, the script should emit no error and continue without requiring a clean working tree. I just now verified that this works as intended in this case.

File: tools/release/backport_pr.sh (hunk 5 above)
Comment (gnossen) [reply tools/release/backport_pr.sh]: Great catch! I should have gone back and checked on this. This would have broken the merged case. This made me realize that I didn't actually test out the old "already merged" case.

I've gone back and fixed a couple of other problems with that case. It should be good now.

**Comment by gnossen** [comment]
@drfloob Thanks for the review! Even more useful than usual, since there are no tests for this script.

File: tools/release/backport_pr.sh [hunk 6]
```diff
@@ -24,59 +24,104 @@ ensure_command () {
   fi
 }
 
+display_usage () {
+  cat << EOF >/dev/stderr
+USAGE: $0 PR_ID GITHUB_USER BACKPORT_BRANCHES REVIEWERS [-c PER_BACKPORT_COMMAND]
```
Comment (drfloob) [inline tools/release/backport_pr.sh:29]: For consistency, replace PR_ID with PR_NUMBER throughout. The variable used in the script is $PR_NUMBER. Or alternatively, change the variable to $PR_ID.

File: tools/release/backport_pr.sh [hunk 7]
```diff
@@ -24,59 +24,104 @@ ensure_command () {
...
+   PR_ID: The ID of the PR to be backported.
+   GITHUB_USER: Your GitHub username.
+   BACKPORT_BRANCHES: A space-separated list of branches to which the
```
Comment (drfloob) [inline tools/release/backport_pr.sh:32]: formatting nit: the line break here is inconsistent with the rest of the file, a few more words could fit.

File: tools/release/backport_pr.sh [hunk 8]
```diff
@@ -24,59 +24,104 @@ ensure_command () {
...
+     source PR will be backported.
+   REVIEWERS: A comma-separated list of users to add as both reviewer and assignee. >/dev/stderr
```
Comment (drfloob) [inline tools/release/backport_pr.sh:34]: There's an extra `>/dev/stderr` at the end here.

**Review by drfloob — APPROVED** [review]
Looks good! Approval given, there are just a few more documentation issues to clean up.

File: tools/release/backport_pr.sh (hunk 8 above)
Comment (gnossen) [reply tools/release/backport_pr.sh]: 🤦
